### ISSUE TITLE
Latest version of Opal (with whitequark/parser ruby25) does not compile Asciidoctor

### ISSUE THREAD

**Issue by Mogztter** [asciidoctor/asciidoctor.js#424]
The build is really slow, then the following exception is thrown:

```
/Users/guillaumegrossetie/dev/asciidoctor.js/node_modules/opal-compiler/src/opal-builder.js:2059
          } else { throw $err; }
                   ^
SyntaxError: unexpected token tIDENTIFIER
    at $Compiler.$$parse [as $parse] (/Users/guillaumegrossetie/dev/asciidoctor.js/node_modules/opal-compiler/src/opal-builder.js:28402:20)
    at $Compiler.$$compile [as $compile] (/Users/guillaumegrossetie/dev/asciidoctor.js/node_modules/opal-compiler/src/opal-builder.js:28381:14)
    at $RubyProcessor.$$compiled [as $compiled] (/Users/guillaumegrossetie/dev/asciidoctor.js/node_modules/opal-compiler/src/opal-builder.js:29860:268)
    at $RubyProcessor.$$requires [as $requires] (/Users/guillaumegrossetie/dev/asciidoctor.js/node_modules/opal-compiler/src/opal-builder.js:29875:23)
    at $Builder.$$process_require [as $process_require] (/Users/guillaumegrossetie/dev/asciidoctor.js/node_modules/opal-compiler/src/opal-builder.js:2194:57)
    at TMP_29 (/Users/guillaumegrossetie/dev/asciidoctor.js/node_modules/opal-compiler/src/opal-builder.js:2213:23)
    at Object.Opal.yield1 (/Users/guillaumegrossetie/dev/asciidoctor.js/node_modules/opal-compiler/node_modules/opal-runtime/src/opal.js:1388:14)
    at Array.$$collect [as $collect] (/Users/guillaumegrossetie/dev/asciidoctor.js/node_modules/opal-compiler/node_modules/opal-runtime/src/opal.js:11897:26)
    at Object.Opal.send (/Users/guillaumegrossetie/dev/asciidoctor.js/node_modules/opal-compiler/node_modules/opal-runtime/src/opal.js:1605:19)
    at Array.alias [as $map] (/Users/guillaumegrossetie/dev/asciidoctor.js/node_modules/opal-compiler/node_modules/opal-runtime/src/opal.js:1780:19)
```

To reproduce this issue, you will need to clone https://github.com/mogztter/opal-node-compiler and https://github.com/mogztter/opal-node-runtime

```
yarn add ../opal-node-runtime
yarn add ../opal-node-compiler --dev
yarn run clean
yarn run build
```

**Comment by Mogztter** [comment]
Apparently, the error is located in `asciidoctor/js/rx`:
 
```
$Diagnostic {
  '$$id': 572922,
  level: 'error',
  reason: 'unexpected_token',
  arguments:
   $Hash {
     '$$smap': { token: 'tIDENTIFIER' },
     '$$map': {},
     '$$keys': [ 'token' ],
     '$$none': undefined,
     '$$proc': undefined },
  location:
   $Range {
     '$$id': 572868,
     source_buffer:
      $Buffer {
        '$$id': 553982,
        name: 'asciidoctor/js/rx',
        source: [String: 'module Asciidoctor\n  # Unicode property aliases (P_<name> constant is equivalent to \\p{<name>})\n  P_L = \'A-Za-zªµºÀ-ÖØ-öø-ˁˆ-ˑˠ-ˤˬˮͰ-ʹͶͷͺ-ͽͿΆΈ-ΊΌΎ-ΡΣ-ϵϷ-ҁҊ-ԯԱ-Ֆՙա-ևא-תװ-ײؠ-يٮٯٱ-ۓەۥۦۮۯۺ-ۼۿܐܒ-ܯݍ-ޥޱߊ-ߪߴߵߺࠀ-ࠕࠚࠤࠨࡀ-ࡘࢠ-ࢴऄ-हऽॐक़-ॡॱ-ঀঅ-ঌএঐও-নপ-রলশ-হঽৎড়ঢ়য়-ৡৰৱਅ-ਊਏਐਓ-ਨਪ-ਰਲਲ਼ਵਸ਼ਸਹਖ਼-ੜਫ਼ੲ-ੴઅ-ઍએ-ઑઓ-નપ-રલળવ-હઽૐૠૡૹଅ-ଌଏଐଓ-ନପ-ରଲଳଵ-ହଽଡ଼ଢ଼ୟ-ୡୱஃஅ-ஊஎ-ஐஒ-கஙசஜஞடணதந-பம-ஹௐఅ-ఌఎ-ఐఒ-నప-హఽౘ-ౚౠౡಅ-ಌಎ-ಐಒ-ನಪ-ಳವ-ಹಽೞೠೡೱೲഅ-ഌഎ-ഐഒ-ഺഽൎൟ-ൡൺ-ൿඅ-ඖක-නඳ-රලව-ෆก-ะาำเ-ๆກຂຄງຈຊຍດ-ທນ-ຟມ-ຣລວສຫອ-ະາຳຽເ-ໄໆໜ-ໟༀཀ-ཇཉ-ཬྈ-ྌက-ဪဿၐ-ၕၚ-ၝၡၥၦၮ-ၰၵ-ႁႎႠ-ჅჇჍა-ჺჼ-ቈቊ-ቍቐ-ቖቘቚ-ቝበ-ኈኊ-ኍነ-ኰኲ-ኵኸ-ኾዀዂ-ዅወ-ዖዘ-ጐጒ-ጕጘ-ፚᎀ-ᎏᎠ-Ᏽᏸ-ᏽᐁ-ᙬᙯ-ᙿᚁ-ᚚᚠ-ᛪᛱ-ᛸᜀ-ᜌᜎ-ᜑᜠ-ᜱᝀ-ᝑᝠ-ᝬᝮ-ᝰក-ឳៗៜᠠ-ᡷᢀ-ᢨᢪᢰ-ᣵᤀ-ᤞᥐ-ᥭᥰ-ᥴᦀ-ᦫᦰ-ᧉᨀ-ᨖᨠ-ᩔᪧᬅ-ᬳᭅ-ᭋᮃ-ᮠᮮᮯᮺ-ᯥᰀ-ᰣᱍ-ᱏᱚ-ᱽᳩ-ᳬᳮ-ᳱᳵᳶᴀ-ᶿḀ-ἕἘ-Ἕἠ-ὅὈ-Ὅὐ-ὗὙὛὝὟ-ώᾀ-ᾴᾶ-ᾼιῂ-ῄῆ-ῌῐ-ΐῖ-Ίῠ-Ῥῲ-ῴῶ-ῼⁱⁿₐ-ₜℂℇℊ-ℓℕℙ-ℝℤΩℨK-ℭℯ-ℹℼ-ℿⅅ-ⅉⅎↃↄⰀ-Ⱞⰰ-ⱞⱠ-ⳤⳫ-ⳮⳲⳳⴀ-ⴥⴧⴭⴰ-ⵧⵯⶀ-ⶖⶠ-ⶦⶨ-ⶮⶰ-ⶶⶸ-ⶾⷀ-ⷆⷈ-ⷎⷐ-ⷖⷘ-ⷞⸯ々〆〱-〵〻〼ぁ-ゖゝ-ゟァ-ヺー-ヿㄅ-ㄭㄱ-ㆎㆠ-ㆺㇰ-ㇿ㐀-䶵一-鿕ꀀ-ꒌꓐ-ꓽꔀ-ꘌꘐ-ꘟꘪꘫꙀ-ꙮꙿ-ꚝꚠ-ꛥꜗ-ꜟꜢ-ꞈꞋ-ꞭꞰ-ꞷꟷ-ꠁꠃ-ꠅꠇ-ꠊꠌ-ꠢꡀ-ꡳꢂ-ꢳꣲ-ꣷꣻꣽꤊ-ꤥꤰ-ꥆꥠ-ꥼꦄ-ꦲꧏꧠ-ꧤꧦ-ꧯꧺ-ꧾꨀ-ꨨꩀ-ꩂꩄ-ꩋꩠ-ꩶꩺꩾ-ꪯꪱꪵꪶꪹ-ꪽꫀꫂꫛ-ꫝꫠ-ꫪꫲ-ꫴꬁ-ꬆꬉ-ꬎꬑ-ꬖꬠ-ꬦꬨ-ꬮꬰ-ꭚꭜ-ꭥꭰ-ꯢ가-힣ힰ-ퟆퟋ-ퟻ豈-舘並-龎ﬀ-ﬆﬓ-ﬗיִײַ-ﬨשׁ-זּטּ-לּמּנּסּףּפּצּ-ﮱﯓ-ﴽﵐ-ﶏﶒ-ﷇﷰ-ﷻﹰ-ﹴﹶ-ﻼＡ-Ｚａ-ｚｦ-ﾾￂ-ￇￊ-ￏￒ-ￗￚ-ￜ\'\n  P_Nl = \'ᛮ-ᛰⅠ-ↂↅ-ↈ〇〡-〩〸-〺ꛦ-ꛯ\'\n  P_Nd = \'0-9٠-٩۰-۹߀-߉०-९০-৯੦-੯૦-૯୦-୯௦-௯౦-౯೦-೯൦-൯෦-෯๐-๙໐-໙༠-༩၀-၉႐-႙០-៩᠐-᠙᥆-᥏᧐-᧙᪀-᪉᪐-᪙᭐-᭙᮰-᮹᱀-᱉᱐-᱙꘠-꘩꣐-꣙꤀-꤉꧐-꧙꧰-꧹꩐-꩙꯰-꯹０-９\'\n  P_Pc = \'_‿⁀⁔︳︴﹍-﹏＿\'\n\n  # Asciidoctor regular expression character classes and groups\n  CC_ALPHA = %(#{P_L}#{P_Nl}) # NOTE Chinese numbers are included in the Alpha properties block\n  CG_ALPHA = %([#{CC_ALPHA}])\n  CC_ALNUM = %(#{CC_ALPHA}#{P_Nd})\n  CG_ALNUM = %([#{CC_ALNUM}])\n  CC_WORD = %(#{CC_ALNUM}#{P_Pc}) # FIXME technically includes M too, but seems superfluous\n  CG_WORD = %([#{CC_WORD}])\n  CG_BLANK = \'[ \\\\t]\' # in AsciiDoc, we only consider space or tab when looking for a blank character\n  CC_EOL = \'(?=\\\\n|$)\' # $ matches end of string (not line) unless multiline flag is set\n  CG_GRAPH = \'[^\\s\\x00-\\x1F\\x7F]\' # non-blank character (equivalent to [^\\p{Z}\\p{C}])\n  CC_ALL = \'[\\s\\S]\' # any character, including newlines\nend\n'],
        first_line: 1,
        lines: [Object],
        line_begins: [Object],
        slice_source: [Object],
        line_for_position: [Object],
        column_for_position: [Object] },
     begin_pos: 1031,
     end_pos: 1037 },
  highlights: [] }
```

**Comment by iliabylich** [comment]
That's weird. parser can parse this file and Opal can compile it to JS. It looks like the error comes from the transpiled parser that acts a bit differently because of the Opal itself.

**Comment by Mogztter** [comment]
Indeed.
But there's a bigger issue, if I replace this code with escape code (cherry-pick https://github.com/asciidoctor/asciidoctor.js/pull/395/commits/f5d9adacb3f408e80536756a1a03a50659051117) then the compilation never ends.

Here's the cpu profile:

![bottom_up](https://user-images.githubusercontent.com/333276/36075899-468bb53e-0f55-11e8-9fdb-5c4ad31322a2.jpeg)

![top_calls](https://user-images.githubusercontent.com/333276/36075900-46ac5064-0f55-11e8-8fac-22c6b18ff13d.jpeg)

![loop](https://user-images.githubusercontent.com/333276/36075901-46cb3f92-0f55-11e8-87b9-a4b7bcc0b1ed.jpeg)

It looks like the compiler/parser is stuck on `unpack` -> `processChunk`. 
But I'm wondering why the previous version worked without `String#unpack`... 🤔

**Comment by iliabylich** [comment]
Previous version had a "fake" unpack method that supported only `"U*"` format string (the one that is used by the `parser`). Current implementation is more compliant with MRI specification (and thus has more potential bugs).

May I ask you why do use `opal-parser`? Can you compile your code with ruby version of Opal? It's faster and it's based on Ruby, so both `parser` and Opal work as they are supposed to work. Moreover, we don't have any tests for `opal-parser`, so some functionality can be broken and catching/fixing such bugs can be a nightmare.

**Comment by Mogztter** [comment]
> Previous version had a "fake" unpack method that supported only "U*" format string (the one that is used by the parser). Current implementation is more compliant with MRI specification (and thus has more potential bugs).

Aha, that's why! Thanks for pointing that out.
I will try to switch to the previous implementation.

> May I ask you why do use opal-parser? 

I'm using `opal-builder` (in JavaScript): https://github.com/opal/opal/blob/v0.11.0/lib/opal/builder.rb
But maybe behind the scene I'm using `opal-parser` ?

>  Can you compile your code with ruby version of Opal? It's faster and it's based on Ruby, so both parser and Opal work as they are supposed to work. 

Yes I can, but the goal is to compile Ruby code to JavaScript using JavaScript.
To be honest it's actually pretty impressive that Opal can compile a compiler 😉 
It sure has some drawbacks but on the bright side the project is easier to build for a JavaScript developer that don't know the Ruby ecosystem. And I think it also benefit Opal by finding weird bugs 🐞 

> Moreover, we don't have any tests for opal-parser, so some functionality can be broken and catching/fixing such bugs can be a nightmare.

It's true, as far as I know the only (smoke) test is https://github.com/opal/opal/blob/3392ee2a35a2d5a42ed5fde47992c564a0981511/test/nodejs/test_opal_builder.rb
Maybe we could create a test runner that use Opal Builder in JavaScript ?

**Comment by Mogztter** [comment]
> Aha, that's why! Thanks for pointing that out.
I will try to switch to the previous implementation.

@iliabylich It's working with the previous implementation, I will investigate 👀

**Comment by iliabylich** [comment]
Could you find which strings gets unpacked differently by Opal and MRI? Maybe I can help you with debugging.

**Comment by Mogztter** [comment]
I think it's more a performance issue on huge String.
Basically `unpack` is called on a large Ruby file and the String contains more than 50K characters.

Previously we were doing the following for `C*`:

**previously.js**
```js
var result = [];
for (i = 0; i < string.length; i++) {
  result.push(string.charCodeAt(i));
}
return result;
```

For each element we are calling `charCodeAt`.
Now we are doing the following:

**now.js**
```js
var buffer = utf16LEToBytes(self); // fast

var result = []
while (buffer.length > 0) {
  var chunk = buffer.slice(0, 1);
  buffer = buffer.slice(1, buffer.length);
  result = result.concat(chunk);
}
```

So for each element we are calling `slice` twice. The first one could be optimized to `var chunk = buffer[0]` but since we are reading the whole buffer, I think we could just return the buffer.

**optimized.js**
```js
var buffer = utf16LEToBytes(self); // fast
if (count === Infinity && (directive === 'C' || directive === 'c' || directive === 'A' || directive === 'a')) {
  return buffer;
}
```

@iliabylich What do you think ?

**Comment by iliabylich** [comment]
Makes sense. Usually nobody uses weird format strings like `C3lUL>ua*`, in most cases it's a letter and a star, so we can optimize happy paths. One thing to mention: `C` and `c` are different. `C` returns bytes, `c` returns `byte >= 128 ? 255 - byte : byte`. `a` and `A` are also different (basically, pretty much all formats are different, there are only 4 aliases).

I'd add a separate mapping `optimized = { "pattern": function(buffer) {}, ... }` and check for existence of optimized version in the beginning of `pack/unpack` methods for fast `return`.

Your code (that simply returns a buffer) is a fast implementation for `C*` (its rule **is** an identity function if you have a buffer of bytes). Other formats (if you need them) like `c*`, `a*` and `A*` can be easily implemented by replicating corresponding chunk reeders + handlers in the `readChunk` and `handlers` mappings. 

Could you send a PR? If not I can do it myself but a bit later.

**Comment by Mogztter** [comment]
> Could you send a PR? If not I can do it myself but a bit later.

Sure, I'm working on it 😄

**Comment by iliabylich** [comment]
Could you test it on the latest Opal master with https://github.com/opal/opal/pull/1766 merged please?

**Comment by Mogztter** [comment]
Yes sure, thanks for the merge :+1:

**Comment by Mogztter** [comment]
Resolved by opal/opal#1766, thanks @iliabylich  :+1:
